Save Python refcount bump on each arg in maybe_handle_torch_function

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164625
* #164624

Pybind's API entails a small unnecessary overhead when working with args. (Similarly, we should probably be using vectorcall, but that's a bigger change for both us and pybind11.)

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

swolchok committed:
Save Python refcount bump on each arg in maybe_handle_torch_function

Pybind's API entails a small unnecessary overhead when working with args. (Similarly, we should probably be using vectorcall, but that's a bigger change for both us and pybind11.)

[ghstack-poisoned]

Skylion007 commented:
We have vector call on the roadmap for pybind11 and would welcome the contribution upstream there.

albanD commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -874,10 +874,19 @@ std::optional<py::object> _maybe_handle_torch_function(
+    // profiling and came here to fix it. For tuples in particular,
+    // PyTuple_GET_ITEM returns a borrowed reference, so no counting
+    // overhead.
+    auto* const args_i_ptr = PyTuple_GET_ITEM(args_ptr, i);
```
Please use PyTuple_GetItem for proper bound checking.

swolchok replied:
we know that the argument is in range because `i` is iterating over `args.size()`. Do you mean that we should ignore this and bounds check anyway?

swolchok replied:
> if we consider this to be perf-critical, I'm ok with skipping the bound check.

I don't think it's perf-critical, just noticeable. I'll change it to PyTuple_GetItem and leave a note that the bounds check could be skipped if anyone comes back and looks at the code because of perf.

albanD commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -874,10 +874,19 @@ std::optional<py::object> _maybe_handle_torch_function(
+    // py::handle, so args[i].ptr() would cause a reference count
+    // round trip. This has enough overhead that I noticed it while
+    // profiling and came here to fix it. For tuples in particular,
+    // PyTuple_GET_ITEM returns a borrowed reference, so no counting
```
Please add a check that this is actually a tuple

albanD reviewed (approved):
Two small nits, bu SGTM

swolchok committed:
add static_assert to clarify we have a statically-known tuple. Rephrase comment to avoid implying we don't know we have tuples. on "Save Python refcount bump on each arg in maybe_handle_torch_function"


Pybind's API entails a small unnecessary overhead when working with args. (Similarly, we should probably be using vectorcall, but that's a bigger change for both us and pybind11.)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

albanD reviewed (approved):
Accepting with the extra assert.

swolchok committed:
use PyTuple_GetItem on "Save Python refcount bump on each arg in maybe_handle_torch_function"


Pybind's API entails a small unnecessary overhead when working with args. (Similarly, we should probably be using vectorcall, but that's a bigger change for both us and pybind11.)

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]
